packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -943,10 +943,50 @@ class RenderObjectToWidgetElement<T extends RenderObject> extends RootRenderObje
   }
 }
 
+/// Base class that provide necessary information to register a ServiceExtension.
+abstract class ExtensionRegistrarBase {
+  /// Name to registers a service extension method with.(full name
+  /// "ext.flutter.name").
+  String get name;
+
+  /// Callback called when the extension method is called.
+  ServiceExtensionCallback get callback;
+}
+
+/// Binding for service extensions to register.
+///
+/// There might be multiple optional ServiceExtensions in practice.
+/// For example, ServiceExtension used by 'flutter driver' is valid only in that command.
+/// Using [ExtensionBinding.registerExtensionRegistrar], developers can register custom service extensions conditionally,
+/// without inheriting & instancing a [BindingBase] instance.
+mixin ExtensionBinding on BindingBase {
+  /// List to save ExtensionRegistrars.
+  static final List<ExtensionRegistrarBase> _extensionRegistrars = <ExtensionRegistrarBase>[];
+
+  /// Calling [ExtensionBinding.registerExtensionRegistrar] with a registrar will save it in [_extensionRegistrars].
+  /// However, the service extension corresponding to the registrar won't be registered directly.
+  /// It will be actually registered until the singleton [BindingBase]'s instantiation.
```

Comment:
nit: grammar is wrong here. "It will be registered when the [BindingBase] singleton is instantiated" or some such. Or "It won't actually be registered until...".